### ISSUE TITLE
torch.no_grad fails to work when running under torch.autocast with float16 base model + float32 lora

### ISSUE THREAD

**Issue by gblackout** [huggingface/peft#291]
Hi there,

I found a weird bug when working with float16 PretrainedModel (e.g., LLaMA) + float32 lora under the `torch.autocast()` manager, where the `torch.no_grad()` starts to affect the gradients outside the of `no_grad` context, and somehow re-entering the `torch.autocast()` context will solve the issue:
```python

dummy_input = {
    'input_ids': torch.tensor([[1,2,3]]).to('cuda'),
}

with torch.autocast(device_type='cuda', dtype=torch.float16):
    with torch.no_grad():
        y = peft_model(**dummy_input)
        print('Inside autocast, and no_grad:\n', y.logits.requires_grad)

    y = peft_model(**dummy_input)
    print('Inside the same autocast, but outside no_grad:\n', y.logits.requires_grad)
    
with torch.autocast(device_type='cuda', dtype=torch.float16):
    y = peft_model(**dummy_input)
    print('Inside another autocast, but outside no_grad:\n', y.logits.requires_grad)
```

Output:

```
Inside autocast, and no_grad:
 False
Inside the same autocast, but outside no_grad:
 False
Inside another autocast, but outside no_grad:
 True
```

Below is a minimalist example to reproduce the error (I'm using the latest peft from github main branch)

```python
from transformers import LlamaForCausalLM, LlamaConfig
from peft import LoraConfig, get_peft_model
import torch
from copy import deepcopy

# init a toy llama model
config = LlamaConfig(
    bos_token_id = 0,
    eos_token_id = 1,
    hidden_act = "silu",
    hidden_size = 8,
    initializer_range = 0.02,
    intermediate_size = 16,
    max_sequence_length = 2048,
    model_type = "llama",
    num_attention_heads = 2,
    num_hidden_layers = 2,
    pad_token_id = -1,
    rms_norm_eps = 1e-06,
    tie_word_embeddings = False,
    torch_dtype = "float16",
    transformers_version = "4.28.0.dev0",
    use_cache = True,
    vocab_size = 32000
)

llama_model = LlamaForCausalLM(config).to('cuda')
llama_model.half() # simulating the official llama weight in float16

# init a toy lora model
lora_config = LoraConfig(
    r=4,
    lora_alpha=4,
    target_modules=['q_proj', 'v_proj'],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
)

peft_model = get_peft_model(deepcopy(llama_model), lora_config)
peft_model.to('cuda')

dummy_input = {
    'input_ids': torch.tensor([[1,2,3]]).to('cuda'),
}

# first check if the base llama_model is working fine
with torch.autocast(device_type='cuda', dtype=torch.float16):
    with torch.no_grad():
        y = llama_model(**dummy_input)
        print('Inside autocast, and no_grad:\n', y.logits.requires_grad)

    y = llama_model(**dummy_input)
    print('Inside the same autocast, but outside no_grad:\n', y.logits.requires_grad)
    
with torch.autocast(device_type='cuda', dtype=torch.float16):
    y = llama_model(**dummy_input)
    print('Inside another autocast, but outside no_grad:\n', y.logits.requires_grad)



# then check the same setting with peft_model
with torch.autocast(device_type='cuda', dtype=torch.float16):
    with torch.no_grad():
        y = peft_model(**dummy_input)
        print('Inside autocast, and no_grad:\n', y.logits.requires_grad)

    y = peft_model(**dummy_input)
    print('Inside the same autocast, but outside no_grad:\n', y.logits.requires_grad)
    
with torch.autocast(device_type='cuda', dtype=torch.float16):
    y = peft_model(**dummy_input)
    print('Inside another autocast, but outside no_grad:\n', y.logits.requires_grad)
```

**Comment by alexjc** [comment]
This would be awesome to address (fits with #336); if the no_grad mode works reliably in 16-bit and even 8-bit then I can do two forward passes to estimate the gradients manually.
